src/jllvm/materialization/ByteCodeCompileLayer.cpp

```diff
@@ -563,21 +563,71 @@ void codeGenBody(llvm::Function* function, const Code& code, const ClassFile& cl
     std::vector<std::pair<std::uint16_t, PoolIndex<ClassInfo>>> activeHandlers;
 
     llvm::DenseMap<std::uint16_t, llvm::BasicBlock*> basicBlocks;
-
-    std::vector<llvm::BasicBlock::iterator> insertPoints(code.getCode().size());
-
-    std::vector<llvm::Value*> operandStack;
-    operandStack.reserve(code.getMaxStack());
-
+    //basicBlocks.insert({0, llvm::BasicBlock::Create(builder.getContext(), "", function)});
+    // Calculate BasicBlocks
     llvm::ArrayRef<char> current = code.getCode();
     while (!current.empty())
     {
         auto offset = current.data() - code.getCode().data();
+        auto addBasicBlock = [&](std::uint16_t target)
+        {
+            auto [result, inserted] = basicBlocks.insert({target, nullptr});
 
-        if (!builder.GetInsertBlock()->empty())
+            if (inserted)
+            {
+                result->second = llvm::BasicBlock::Create(builder.getContext(), "", function);
+            }
+        };
+        auto opCode = consume<OpCodes>(current);
+        switch(opCode)
         {
-            insertPoints[offset] = builder.GetInsertBlock()->back().getIterator();
+            default:
+            {
+                break;
```

Comment:
The default case here is a bit problematic because JVM bytecode instructions are multi byte and this would lead to false detection and crashes if these happen to have bytes that look like the branch instructions below.

As an example:
```java
public class Test {
    public static int foo()
    {
        return -95;
    }

    public static void main(String[] args) {
        foo();
    }
}
```
crashes here since `-95` happens to be the opcode for `if_icmpeq`. It then tries to read in two bytes after in `consume` below, but there is only a single byte after the `bipush` (which is an `ireturn` https://godbolt.org/z/K3eKc7jxY).

So in reality we'd have to handle all instructions here. Ideally we'd do this in a code reusable way for the loop below as well. That is a bit more complicated tho.

What you could do for the time being maybe is just do bounds checking below? So just make sure that `current.size()` is greater or equal to `sizeof(std::int16_t)`. We can still have false positives this way creating too many basic blocks, but we shouldn't have any crashes or correctness issues this way. 
We can do the nice code reuse as part of a separate issue if you like.